jms/src/main/scala/akka/stream/alpakka/jms/JmsConnector.scala

```diff
@@ -136,33 +138,20 @@ private[jms] trait JmsConnector[S <: JmsSession] {
           if (maxRetries == 0) Future.failed(t)
           else Future.failed(ConnectionRetryException(s"Could not establish connection after $n retries.", t))
         } else {
-          val delay = if (maxed) maxBackoff else waitTime(nextN)
-          if (delay >= maxBackoff) {
-            after(maxBackoff, system.scheduler) {
-              openConnectionWithRetry(startConnection, nextN, maxed = true)
-            }
-          } else {
-            after(delay, system.scheduler) {
-              openConnectionWithRetry(startConnection, nextN)
-            }
+          val delay = waitTime(nextN).min(maxBackoff)
+          after(delay, system.scheduler) {
+            openConnectionWithRetry(startConnection, nextN)
           }
         }
     }(ExecutionContexts.sameThreadExecutionContext)
 
-  private[jms] def openRecoverableConnection(startConnection: Boolean,
-                                             onConnectionFailure: jms.JMSException => Unit): Future[jms.Connection] = {
+  private[jms] def openRecoverableConnection(startConnection: Boolean): Future[jms.Connection] = {
     implicit val system: ActorSystem = ActorMaterializerHelper.downcast(materializer).system
     jmsConnection = openConnectionWithRetry(startConnection).map { connection =>
       connection.setExceptionListener(new jms.ExceptionListener {
-        override def onException(ex: jms.JMSException) = {
-          try {
-            connection.close() // best effort closing the connection.
-          } catch {
-            case _: Throwable =>
-          }
-          jmsSessions = Seq.empty
-
-          onConnectionFailure(ex)
+        override def onException(ex: jms.JMSException): Unit = {
+          Try(connection.close()) // best effort closing the connection.
+          connectionFailedCB.invoke(ex)
```

Comment:
here also, I think we must switch back to the graph stage threads with the asyncCallback. This however changed the amount of reconnects in the test.